File: jpms-java/src/main/java/example/jpms_java/repository/EmployeeRepository.java
package example.jpms_java.repository;

import example.jpms_java.domain.Age;
import example.jpms_java.domain.Salary;
import example.jpms_java.entity.Department_;
import example.jpms_java.entity.Employee;
import example.jpms_java.entity.Employee_;
import example.jpms_java.entity.NameAndSalaryDto;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import org.seasar.doma.jdbc.Config;
import org.seasar.doma.jdbc.criteria.Entityql;
import org.seasar.doma.jdbc.criteria.NativeSql;
import org.seasar.doma.jdbc.criteria.expression.Expressions;
import org.seasar.doma.jdbc.criteria.option.LikeOption;
import org.seasar.doma.jdbc.criteria.statement.StreamMappable;
import org.seasar.doma.jdbc.criteria.tuple.Tuple2;

public class EmployeeRepository {

  private final Entityql entityql;

  private final NativeSql nativeSql;

  public EmployeeRepository(Config config) {
    Objects.requireNonNull(config);
    entityql = new Entityql(config);
    nativeSql = new NativeSql(config);
  }

  public Employee selectById(Integer id) {
    Employee_ e = new Employee_();
    return entityql.from(e).where(c -> c.eq(e.id, id)).fetchOne();
  }

  public List<Employee> selectByAgeRange(Age min, Age max) {
    Employee_ e = new Employee_();
    return entityql
        .from(e)
        .where(
            c -> {
              c.ge(e.age, min);
              c.le(e.age, max);
            })
        .fetch();
  }

  public List<Employee> selectByName(String name) {
    Employee_ e = new Employee_();
    return entityql.from(e).where(c -> c.eq(e.name, name)).fetch();
  }

  public List<Employee> selectByNames(List<String> names) {
    Employee_ e = new Employee_();
    return entityql.from(e).where(c -> c.in(e.name, names)).fetch();
  }

  public List<Employee> selectByAges(List<Age> ages) {
    Employee_ e = new Employee_();
    return entityql.from(e).where(c -> c.in(e.age, ages)).fetch();
  }

  public List<Employee> selectByNotEmptyName(String name) {
    Employee_ e = new Employee_();
    return entityql
        .from(e)
        .where(
            c -> {
              if (name != null && !name.isEmpty()) {
                c.eq(e.name, name);
              }
            })
        .fetch();
  }

  public List<Employee> selectByNameWithPrefixMatching(String prefix) {
    Employee_ e = new Employee_();
    return entityql.from(e).where(c -> c.like(e.name, prefix, LikeOption.prefix())).fetch();
  }

  public List<Employee> selectByNameWithSuffixMatching(String suffix) {
    Employee_ e = new Employee_();
    return entityql.from(e).where(c -> c.like(e.name, suffix, LikeOption.suffix())).fetch();
  }

  public List<Employee> selectByNameWithInfixMatching(String inside) {
    Employee_ e = new Employee_();
    return entityql.from(e).where(c -> c.like(e.name, inside, LikeOption.infix())).fetch();
  }

  public List<Employee> selectByHiredateRange(LocalDateTime from, LocalDateTime to) {
    LocalDate fromDate = from.toLocalDate();
    LocalDate toDate = to.toLocalDate().plusDays(1);
    Employee_ e = new Employee_();
    return entityql
        .from(e)
        .where(
            c -> {
              c.ge(e.hiredate, fromDate);
              c.lt(e.hiredate, toDate);
            })
        .fetch();
  }

  public List<Employee> selectBySalary(Salary salary) {
    Employee_ e = new Employee_();
    return entityql.from(e).where(c -> c.gt(e.salary, salary)).fetch();
  }

  public Optional<Salary> selectSummedSalary() {
    Employee_ e = new Employee_();
    return nativeSql.from(e).select(Expressions.sum(e.salary)).fetchOptional();
  }

  public List<Employee> selectByExample(Employee employee) {
    Employee_ e = new Employee_();
    return entityql.from(e).where(c -> c.eq(e.name, employee.getName())).fetch();
  }

  public List<Employee> selectAll() {
    return select(null, null);
  }

  public Tuple2<List<Employee>, Long> selectAndCount(Integer offset, Integer limit) {
    List<Employee> list = select(offset, limit);
    Employee_ e = new Employee_();
    long count = nativeSql.from(e).select(Expressions.count()).fetchOptional().orElse(0L);
    return new Tuple2<>(list, count);
  }

  public List<Employee> select(Integer offset, Integer limit) {
    Employee_ e = new Employee_();
    return nativeSql.from(e).orderBy(c -> c.asc(e.id)).offset(offset).limit(limit).fetch();
  }

  public <R> R selectByAge(Age age, Function<Stream<Employee>, R> mapper) {
    Employee_ e = new Employee_();
    StreamMappable<Employee> stmt =
        nativeSql.from(e).where(c -> c.gt(e.age, age)).orderBy(c -> c.asc(e.age));
    return stmt.mapStream(mapper);
  }

  public long selectCount() {
    Employee_ e = new Employee_();
    return nativeSql.from(e).select(Expressions.count()).fetchOptional().orElse(0L);
  }

  public List<Employee> selectAllWithAssociation() {
    Employee_ e = new Employee_();
    Department_ d = new Department_();
    return entityql
        .from(e)
        .leftJoin(d, on -> on.eq(e.departmentId, d.id))
        .orderBy(c -> c.asc(e.id))
        .associate(
            e,
            d,
            (employee, department) -> {
              employee.setDepartment(department);
              department.getEmployees().add(employee);
            })
        .fetch();
  }

  public List<Employee> selectByDepartmentName_in(String departmentName) {
    Employee_ e = new Employee_();
    Department_ d = new Department_();
    return entityql
        .from(e)
        .where(
            c ->
                c.in(
                    e.departmentId,
                    c.from(d).where(c2 -> c2.eq(d.name, departmentName)).select(d.id)))
        .fetch();
  }

  public List<Employee> selectByDepartmentName_exists(String departmentName) {
    Employee_ e = new Employee_();
    Department_ d = new Department_();
    return entityql
        .from(e)
        .where(
            c ->
                c.exists(
                    c.from(d)
                        .where(
                            c2 -> {
                              c2.eq(e.departmentId, d.id);
                              c2.eq(d.name, departmentName);
                            })))
        .fetch();
  }

  public List<Employee> selectByDepartmentName_join(String departmentName) {
    Employee_ e = new Employee_();
    Department_ d = new Department_();
    return entityql
        .from(e)
        .innerJoin(d, on -> on.eq(e.departmentId, d.id))
        .where(c -> c.eq(d.name, departmentName))
        .fetch();
  }

  public List<Employee> selectNameAndSalaryAsEntityList() {
    Employee_ e = new Employee_();
    return entityql.from(e).selectTo(e, e.name, e.salary).fetch();
  }

  public List<Tuple2<String, Salary>> selectNameAndSalaryAsTuple2List() {
    Employee_ e = new Employee_();
    return nativeSql.from(e).select(e.name, e.salary).fetch();
  }

  public List<NameAndSalaryDto> selectNameAndSalaryAsNameAndSalaryDtoList() {
    Employee_ e = new Employee_();
    return nativeSql.from(e).select(e.name, e.salary).stream()
        .map(tuple -> new NameAndSalaryDto(tuple.getItem1(), tuple.getItem2()))
        .collect(Collectors.toList());
  }

  public void insert(Employee employee) {
    Employee_ e = new Employee_();
    entityql.insert(e, employee).execute();
  }

  public void insertByNativeSql(Employee employee) {
    Employee_ e = new Employee_();
    nativeSql
        .insert(e)
        .values(
            c -> {
              c.value(e.id, employee.getId());
              c.value(e.name, employee.getName());
              c.value(e.age, employee.getAge());
              c.value(e.departmentId, employee.getDepartmentId());
              c.value(e.hiredate, employee.getHiredate());
              c.value(e.jobType, employee.getJobType());
              c.value(e.salary, employee.getSalary());
              c.value(e.insertTimestamp, employee.getInsertTimestamp());
              c.value(e.updateTimestamp, employee.getUpdateTimestamp());
              c.value(e.version, employee.getVersion());
            })
        .execute();
  }

  public void update(Employee employee) {
    Employee_ e = new Employee_();
    entityql.update(e, employee).execute();
  }

  public void updateByNativeSql(Employee employee) {
    Employee_ e = new Employee_();
    nativeSql
        .update(e)
        .set(
            c -> {
              c.value(e.name, employee.getName());
              c.value(e.age, employee.getAge());
              c.value(e.departmentId, employee.getDepartmentId());
              c.value(e.hiredate, employee.getHiredate());
              c.value(e.jobType, employee.getJobType());
              c.value(e.salary, employee.getSalary());
              c.value(e.updateTimestamp, employee.getUpdateTimestamp());
              c.value(e.version, employee.getVersion());
            })
        .where(c -> c.eq(e.id, employee.getId()))
        .execute();
  }

  public void delete(Employee employee) {
    Employee_ e = new Employee_();
    entityql.delete(e, employee).execute();
  }

  public void deleteByNativeSql(Employee employee) {
    Employee_ e = new Employee_();
    nativeSql.delete(e).where(c -> c.eq(e.id, employee.getId())).execute();
  }

  public void batchInsert(List<Employee> employees) {
    Employee_ e = new Employee_();
    entityql.insert(e, employees).execute();
  }

  public void batchUpdate(List<Employee> employees) {
    Employee_ e = new Employee_();
    entityql.update(e, employees).execute();
  }

  public void batchDelete(List<Employee> employees) {
    Employee_ e = new Employee_();
    entityql.delete(e, employees).execute();
  }
}

File: dsl-style-kotlin/src/main/kotlin/example/dsl_style_kotlin/repository/EmployeeRepository.kt
package example.dsl_style_kotlin.repository

import example.dsl_style_kotlin.domain.Age
import example.dsl_style_kotlin.domain.Salary
import example.dsl_style_kotlin.entity.Department_
import example.dsl_style_kotlin.entity.Employee
import example.dsl_style_kotlin.entity.Employee_
import org.seasar.doma.jdbc.Config
import org.seasar.doma.jdbc.criteria.KEntityql
import org.seasar.doma.jdbc.criteria.KNativeSql
import org.seasar.doma.jdbc.criteria.expression.KExpressions
import org.seasar.doma.jdbc.criteria.option.LikeOption
import java.time.LocalDateTime

class EmployeeRepository(config: Config) {

    private val entityql: KEntityql = KEntityql(config)
    private val nativeSql: KNativeSql = KNativeSql(config)

    fun selectById(id: Int): Employee {
        val e = Employee_()
        return entityql.from(e).where { eq(e.id, id) }.fetchOne()
    }

    fun selectByAgeRange(min: Age?, max: Age?): List<Employee> {
        val e = Employee_()
        return entityql
            .from(e)
            .where {
                ge(e.age, min)
                le(e.age, max)
            }
            .fetch()
    }

    fun selectByName(name: String?): List<Employee> {
        val e = Employee_()
        return entityql.from(e).where { eq(e.name, name) }.fetch()
    }

    fun selectByNames(names: List<String>?): List<Employee> {
        val e = Employee_()
        return entityql.from(e).where { `in`(e.name, names) }.fetch()
    }

    fun selectByAges(ages: List<Age>?): List<Employee> {
        val e = Employee_()
        return entityql.from(e).where { `in`(e.age, ages) }.fetch()
    }

    fun selectByNotEmptyName(name: String?): List<Employee> {
        val e = Employee_()
        return entityql
            .from(e)
            .where {
                if (name != null && name.isNotEmpty()) {
                    eq(e.name, name)
                }
            }
            .fetch()
    }

    fun selectByNameWithPrefixMatching(prefix: String?): List<Employee> {
        val e = Employee_()
        return entityql.from(e).where { like(e.name, prefix, LikeOption.prefix()) }.fetch()
    }

    fun selectByNameWithSuffixMatching(suffix: String?): List<Employee> {
        val e = Employee_()
        return entityql.from(e).where { like(e.name, suffix, LikeOption.suffix()) }.fetch()
    }

    fun selectByNameWithInfixMatching(inside: String?): List<Employee> {
        val e = Employee_()
        return entityql.from(e).where { like(e.name, inside, LikeOption.infix()) }.fetch()
    }

    fun selectByHiredateRange(from: LocalDateTime, to: LocalDateTime): List<Employee> {
        val fromDate = from.toLocalDate()
        val toDate = to.toLocalDate().plusDays(1)
        val e = Employee_()
        return entityql
            .from(e)
            .where {
                ge(e.hiredate, fromDate)
                lt(e.hiredate, toDate)
            }
            .fetch()
    }

    fun selectBySalary(salary: Salary): List<Employee> {
        val e = Employee_()
        return entityql.from(e).where { gt(e.salary, salary) }.fetch()
    }

    fun selectSummedSalary(): Salary? {
        val e = Employee_()
        return nativeSql.from(e).select(KExpressions.sum(e.salary)).fetchOneOrNull()
    }

    fun selectByExample(employee: Employee): List<Employee> {
        val e = Employee_()
        return entityql.from(e).where { eq(e.name, employee.name) }.fetch()
    }

    fun selectAll(): List<Employee> {
        return select(null, null)
    }

    fun selectAndCount(offset: Int?, limit: Int?): Pair<List<Employee>, Long> {
        val list = select(offset, limit)
        val e = Employee_()
        val count = nativeSql.from(e).select(KExpressions.count()).fetchOneOrNull() ?: 0L
        return list to count
    }

    fun select(offset: Int?, limit: Int?): List<Employee> {
        val e = Employee_()
        return nativeSql.from(e).orderBy { asc(e.id) }.offset(offset).limit(limit).fetch()
    }

    fun <R> selectByAge(age: Age, mapper: (Sequence<Employee>) -> R): R {
        val e = Employee_()
        val stmt = nativeSql.from(e).where { gt(e.age, age) }.orderBy { asc(e.age) }
        return stmt.mapSequence(mapper)
    }

    fun selectCount(): Long {
        val e = Employee_()
        return nativeSql.from(e).select(KExpressions.count()).fetchOneOrNull() ?: 0L
    }

    fun selectAllWithAssociation(): List<Employee> {
        val e = Employee_()
        val d = Department_()
        return entityql
            .from(e)
            .leftJoin(d) { eq(e.departmentId, d.id) }
            .orderBy { asc(e.id) }
            .associate(e, d) { employee, department -> employee.department = department }
            .fetch()
    }

    fun insert(employee: Employee) {
        val e = Employee_()
        entityql.insert(e, employee).execute()
    }

    fun insertByNativeSql(employee: Employee): Int {
        val e = Employee_()
        return nativeSql
            .insert(e)
            .values {
                value(e.id, employee.id)
                value(e.name, employee.name)
                value(e.age, employee.age)
                value(e.departmentId, employee.departmentId)
                value(e.hiredate, employee.hiredate)
                value(e.jobType, employee.jobType)
                value(e.salary, employee.salary)
                value(e.insertTimestamp, employee.insertTimestamp)
                value(e.updateTimestamp, employee.updateTimestamp)
                value(e.version, employee.version)
            }
            .execute()
    }

    fun update(employee: Employee) {
        val e = Employee_()
        entityql.update(e, employee).execute()
    }

    fun updateByNativeSql(employee: Employee): Int {
        val e = Employee_()
        return nativeSql
            .update(e)
            .set {
                value(e.name, employee.name)
                value(e.age, employee.age)
                value(e.departmentId, employee.departmentId)
                value(e.hiredate, employee.hiredate)
                value(e.jobType, employee.jobType)
                value(e.salary, employee.salary)
                value(e.updateTimestamp, employee.updateTimestamp)
                value(e.version, employee.version)
            }
            .where { eq(e.id, employee.id) }
            .execute()
    }

    fun delete(employee: Employee) {
        val e = Employee_()
        entityql.delete(e, employee).execute()
    }

    fun deleteByNativeSql(employee: Employee): Int {
        val e = Employee_()
        return nativeSql.delete(e).where { eq(e.id, employee.id) }.execute()
    }

    fun batchInsert(employees: List<Employee>) {
        val e = Employee_()
        val stmt = entityql.insert(e, employees)
        stmt.execute()
    }

    fun batchUpdate(employees: List<Employee>) {
        val e = Employee_()
        entityql.update(e, employees).execute()
    }

    fun batchDelete(employees: List<Employee>) {
        val e = Employee_()
        entityql.delete(e, employees).execute()
    }
}
